### PR TITLE
ISPN-2634 Implement cross-site replication based on IRAC

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-2634

### CODE REVIEW

**Comment by tristantarrant** [comment]
I would really like to see a server integration test

**Comment by pruivo** [comment]
@tristantarrant added! 
Not sure if it is the correct way to do it... but it is working.

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+      Object key = cleanupTask.key;
+      IracMetadata metadata = iracVersionGenerator.getTombstone(key);
+      if (metadata == null) {
+         return null;
+      }
+      cleanupTask.tombstone = metadata;
+      return commandsFactory.buildIracUpdateKeyCommand(key, null, null, metadata);
+   }
+
+   private CompletionStage<InternalCacheEntry<Object, Object>> fetchEntry(Object key, int segmentId) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:394]: This method should be replaced by a reference to `EntryLoader#loadAndStoreInDataContainer` such as https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/container/entries/VersionedRepeatableReadEntry.java#L98

Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:394]: It handles expiration, notifications and ordering if passivation is enabled. The data container lock is not used for this purpose anymore.

File: core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java [hunk 2]
```diff
@@ -127,58 +124,66 @@ private Object handleSingleKeyWriteCommand(InvocationContext ctx, DataWriteComma
...
+      int segment = dataWriteCommand.getSegment();
+      DistributionInfo dInfo = clusteringDependentLogic.getCacheTopology().getSegmentDistribution(segment);
+      if (dInfo.isWriteOwner()) {
+         //all owners has to keep track of the updates even if the primary owner is the only one who sends it.
+         iracManager.trackUpdatedKey(dataWriteCommand.getKey(), dataWriteCommand.getCommandInvocationId());
+      }
+      if (dInfo.isPrimary()) { //primary owner sends for sync backups
+         CacheEntry<?,?> entry = ctx.lookupEntry(dataWriteCommand.getKey());
+         WriteCommand crossSiteCommand = createCommandForXSite(entry, segment, dataWriteCommand.getFlagsBitSet());
+         return backupSender.backupWrite(crossSiteCommand, dataWriteCommand).thenReturn(ctx, dataWriteCommand, rv);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java:143]: Sorry I can't quite follow, but what this is for? Isn't the irac manager handling the replication of values and the prior if stores the keys in backup owners as well, I guess miss what part this does.

Maybe `hasEnabledBackups` should be `hasSyncEnabledBackups` ? at EmptyConstructorNamedCacheFactory

```java
         return configuration.sites().hasEnabledBackups() ?
                new BackupSenderImpl() :
                NoOpBackupSender.getInstance();
```

File: core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+   private void resolveConflict(CacheEntry<?, ?> entry, DataWriteCommand command, IracMetadata localMetadata,
+         IracMetadata remoteMetadata) {
+      if (isTraceEnabled()) {
+         getLog().tracef("[IRAC] Conflict found between local and remote metadata: %s and %s", localMetadata,
+               remoteMetadata);
+      }
+      //same site? conflict?
+      assert !localMetadata.getSite().equals(remoteMetadata.getSite());
+      if (localMetadata.getSite().compareTo(remoteMetadata.getSite()) < 0) {
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java:89]: yes. 
I'm planning to add some SPI in the future so the users can inject their own merge mechanism when conflicts are detected.

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+      return vectorClock.equals(version.vectorClock);
+   }
+
+   @Override
+   public int hashCode() {
+      return vectorClock.hashCode();
+   }
+
+   @ProtoField(number = 1, collectionImplementation = ArrayList.class)
+   List<MapEntry> entries() {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:112]: If we place this below the `@ProtoFactory` method it makes the marshalling logic easier to read as it's all in one place.

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    * @return A {@link InequalVersionComparisonResult} instance with the compare result.
+    */
+   public InequalVersionComparisonResult compareTo(IracEntryVersion other) {
+      VectorClock vectorClock = new VectorClock();
+      this.forEach(vectorClock::setOurs);
+      other.forEach(vectorClock::setTheirs);
+
+      Merger merger = Merger.NONE;
+      for (VersionCompare v : vectorClock.values()) {
+         merger = merger.accept(v);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:80]: no need. the usual use case is 2 sites and the first compare never generates a conflict.

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+      abstract InequalVersionComparisonResult result();
+   }
+
+   @ProtoTypeId(ProtoStreamTypeIds.IRAC_VERSION_ENTRY)
+   public static class MapEntry {
+
+      @ProtoField(number = 1, type = Type.STRING, required = true)
+      final String site;
+      @ProtoField(number = 2, required = true)
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:234]: `required` should be removed.

File: core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracVersionGenerator.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+   }
+
+   @Override
+   public void updateVersion(int segment, IracEntryVersion remoteVersion) {
+      segmentVersion.merge(segment, remoteVersion.toMap(), this::mergeVectorsFunction);
+   }
+
+   @TopologyChanged
+   public void onTopologyChange(TopologyChangedEvent<?, ?> tce) {
+      topologyId = tce.getNewTopologyId();
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracVersionGenerator.java:70]: :+1:

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+      abstract Merger accept(VersionCompare versions);
+
+      abstract InequalVersionComparisonResult result();
+   }
+
+   @ProtoTypeId(ProtoStreamTypeIds.IRAC_VERSION_ENTRY)
+   public static class MapEntry {
+
+      @ProtoField(number = 1, type = Type.STRING, required = true)
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:232]: type is redundant and `required` should be removed.

File: core/src/main/java/org/infinispan/context/impl/LocalTxInvocationContext.java [hunk 9]
```diff
@@ -64,4 +66,30 @@ public final void remoteLocksAcquired(Collection<Address> nodes) {
    public final Transaction getTransaction() {
       return getCacheTransaction().getTransaction();
    }
+
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/context/impl/LocalTxInvocationContext.java:69]: Additional empty line

File: core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+         IracUtils.discardUpdate(getLog(), entry, command, remoteMetadata);
+         return;
+      }
+      //other site update has priority!
+      setIracMetadata(entry, remoteMetadata);
+      //TODO! this isn't required now but when we allow custom conflict resolution, we need to merge the versions!
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java:95]: Can you create a Jira for this and add the ID to the comment.

File: core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceContextInitializer.java [hunk 11]
```diff
@@ -40,7 +43,12 @@
             SimpleClusteredVersion.class,
             MarshallableUserObject.class,
             PrivateMetadata.class,
-            InDoubtTxInfo.class
+            InDoubtTxInfo.class,
+            MarshallableUserObject.class,
+            IracEntryVersion.class,
+            IracEntryVersion.MapEntry.class,
+            TopologyIracVersion.class,
+            IracMetadata.class
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceContextInitializer.java:51]: I assume `InDoubtTxInfo` will be removed when this is rebased on https://github.com/infinispan/infinispan/pull/8290

File: core/src/main/java/org/infinispan/metadata/impl/PrivateMetadata.java [hunk 12]
```diff
@@ -52,34 +55,59 @@ public static Builder getBuilder(PrivateMetadata metadata) {
     * It allows to do some logic before instantiate a new instance.
     */
    @ProtoFactory
-   static PrivateMetadata protostreamFactory(/*add fields later*/) {
-      return EMPTY;
+   static PrivateMetadata protostreamFactory(IracMetadata iracMetadata) {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/metadata/impl/PrivateMetadata.java:58]: `protoFactory`?

File: core/src/main/java/org/infinispan/metadata/impl/PrivateMetadata.java [hunk 13]
```diff
@@ -21,11 +22,13 @@
    /**
     * A cached empty {@link PrivateMetadata}.
     */
-   private static final PrivateMetadata EMPTY = new PrivateMetadata();
+   private static final PrivateMetadata EMPTY = new PrivateMetadata(null);
 
-   //The following fields will be added in later PRs:
-   //IracVersion => used by IRAC (async xsite new algorithm)
-   //SimpleClusteredVersion => used by Optimistic Transactions
+   private final IracMetadata iracMetadata;
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/metadata/impl/PrivateMetadata.java:27]: We can make this:

```java
@ProtoField(number = 1)
final IracMetadata iracMetadata
```

That way the marshalling info is always at the top of the class and easier to read.

File: core/src/main/java/org/infinispan/xsite/BackupReceiver.java [hunk 14]
```diff
@@ -17,10 +20,51 @@
...
 
-   Cache getCache();
+   Cache<?, ?> getCache();
 
    CompletionStage<Void> handleRemoteCommand(VisitableCommand command, boolean preserveOrder);
 
+   /**
+    * Updates the key with the value from a remote site.
+    * <p>
+    * If a conflict happens, the update can be discarded.
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/xsite/BackupReceiver.java:30]: Nitpick but how about instead of saying "If a conflict happens", we say "If a conflict occurs". The former doesn't sound correct to me.

File: core/src/main/java/org/infinispan/interceptors/impl/AbstractIracLocalSiteInterceptor.java [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+   static RemoteTxInvocationContext asRemoteTxInvocationContext(InvocationContext ctx) {
+      assert !ctx.isOriginLocal();
+      assert ctx.isInTxScope();
+      return (RemoteTxInvocationContext) ctx;
+   }
+
+   abstract boolean isTraceEnabled();
+
+   abstract Log getLog();
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/AbstractIracLocalSiteInterceptor.java:56]: I wasn't thinking about any optimization. It is just to have all the log entries under the same class.

File: core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java [hunk 16]
```diff
@@ -300,6 +319,22 @@ private AsyncInterceptorChain buildInterceptorChain() {
             //Nothing...
       }
 
+      if (cacheMode.isClustered()) {
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/factories/InterceptorChainFactory.java:322]: Can't do that. 
Unfortunately, a cache without IRAC enabled can still receive IRAC updates from the remote site and the `*IracRemoteInterceptor` needs to handle them to discard duplicated or out-of-order updates.

File: core/src/main/java/org/infinispan/xsite/irac/IracUtils.java [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+import org.infinispan.metadata.impl.PrivateMetadata;
+import org.infinispan.util.logging.Log;
+
+/**
+ * Utility methods related to IRAC (async Cross-Site Replication)
+ *
+ * @author Pedro Ruivo
+ * @since 11.0
+ */
+public final class IracUtils {
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/xsite/irac/IracUtils.java:19]: Currently these methods are only ever used by interceptors that extend `AbstractIracLocalSiteInterceptor`. Does it make more sense for these to be simply be defined in `AbstractIracLocalSiteInterceptor` and `AbstractIracRemoteSiteInterceptor`? Or do you envisage these methods being required elsewhere longer term.

File: core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+   public void testWriteAndLoad(Method method) {
+      String key = TestingUtil.k(method);
+      String value = TestingUtil.v(method);
+      IracMetadata metadata = createMetadata();
+
+      cacheStore.write(createEntry(key, value, metadata));
+
+      MarshallableEntry<String, V> loadedMEntry = cacheStore.loadEntry(key);
+
+
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java:61]: Unnecessary whitespace.

File: core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+   protected SerializationContextInitializer getSerializationContextInitializer() {
+      return TestDataSCI.INSTANCE;
+   }
+
+   protected abstract void addPersistence(ConfigurationBuilder builder);
+
+   protected abstract V wrap(String key, String value);
+
+   protected abstract String unwrap(V value);
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java:119]: Can we have the configuration methods above the tests. It makes it easier to read the test configuration etc.

File: core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+      return cm;
+   }
+
+   protected SerializationContextInitializer getSerializationContextInitializer() {
+      return TestDataSCI.INSTANCE;
+   }
+
+   protected abstract void addPersistence(ConfigurationBuilder builder);
+
+   protected abstract V wrap(String key, String value);
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java:117]: The `wrap` and `unwrap` implementations always return the value except in the case of the `IracJpaStoreTest`. We should create a default impl of:

```java
   @Override
   protected String wrap(String key, String value) {
      return value;
   }

   @Override
   protected String unwrap(String value) {
      return value;
   }
```

and override it in `IracJpaStoreTest`.

File: core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+      //noinspection unchecked
+      entryFactory = TestingUtil.extractComponent(cm.getCache(), MarshallableEntryFactory.class);
+      return cm;
+   }
+
+   protected SerializationContextInitializer getSerializationContextInitializer() {
+      return TestDataSCI.INSTANCE;
+   }
+
+   protected abstract void addPersistence(ConfigurationBuilder builder);
```
Comment (ryanemerson) [inline core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java:115]: How about just `configure` as we're passing the `ConfigurationBuilder` so it's possible for implementations to configure any component as required.

File: core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java [hunk 22]
```diff
@@ -0,0 +1,248 @@
...
+      return ctx.isInTxScope() || command.hasAnyFlag(FlagBitSets.IRAC_UPDATE);
+   }
+
+   /**
+    * Visits the {@link WriteCommand} before executing it.
+    * <p>
+    * The primary owner generates a new {@link IracMetadata} and stores it in the {@link WriteCommand}.
+    */
+   private void visitKey(Object key, WriteCommand command) {
+      int segment = getSegment(key);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java:194]: :+1:

File: core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracVersionGenerator.java [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+   }
+
+   private Map<String, TopologyIracVersion> generateNewVectorFunction(Integer s,
+         Map<String, TopologyIracVersion> versions) {
+      if (versions == null) {
+         return Collections.singletonMap(localSite, TopologyIracVersion.newVersion(topologyId));
+      } else {
+         Map<String, TopologyIracVersion> copy = new HashMap<>(versions);
+         copy.compute(localSite, this::incrementVersionFunction);
+         return Collections.unmodifiableMap(copy);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracVersionGenerator.java:111]: :+1:

File: core/src/main/java/org/infinispan/xsite/irac/IracUtils.java [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+
+   private IracUtils() {
+
+   }
+
+   private static void logIracMetadataAssociated(Log log, Object key, IracMetadata metadata) {
+      if (log.isDebugEnabled()) {
+         log.debugf("[IRAC] IracMetadata %s associated with key '%s'", metadata, key);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/IracUtils.java:27]: My idea was to have the versions as debug so we can enable it in runtime. 
I'll change to trace.

File: core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java [hunk 25]
```diff
@@ -0,0 +1,248 @@
...
+      if (getOwnership(segment) != Ownership.PRIMARY) {
+         return;
+      }
+      IracMetadata metadata = iracVersionGenerator.generateNewMetadata(segment);
+      if (log.isDebugEnabled()) {
+         log.debugf("[IRAC] New metadata for key '%s' is %s", key, metadata);
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java:200]: :+1:

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 1 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:394]: damn... The EntryLoader is really hidden xD

Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:394]: hmmm. I can change it later. The `EntryLoader` requires the `InvocationContext` and a `Command` and IRAC doesn't have that.

File: core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java (hunk 2 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java:143]: so, IRAC handles the async xsite backups and the BackupSender handles only the sync xsite backups. 
I'll change the EmptyConstructorNamedCacheFactory

File: core/src/main/java/org/infinispan/metadata/impl/IracMetadata.java [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+
+   public static void writeTo(ObjectOutput output, IracMetadata metadata) throws IOException {
+      metadata.writeTo(output);
+   }
+
+   public static IracMetadata readFrom(ObjectInput in) throws IOException, ClassNotFoundException {
+      return new IracMetadata(in.readUTF(), (IracEntryVersion) in.readObject());
+   }
+
+   @ProtoField(number = 1, type = Type.STRING, required = true)
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/metadata/impl/IracMetadata.java:41]: :+1:

File: core/src/main/java/org/infinispan/commands/irac/IracUpdateKeyCommand.java [hunk 27]
```diff
@@ -0,0 +1,249 @@
...
+
+   public CompletionStage<Void> executeOperation(BackupReceiver receiver) {
+      return operation.execute(receiver);
+   }
+
+   @Override
+   public CompletionStage<?> invokeAsync(ComponentRegistry registry) {
+      //noinspection unchecked
+      Cache<Object, Object> cache = registry.getCache().running();
+      //TODO! create a component for BackupReceiver
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/irac/IracUpdateKeyCommand.java:107]: definitely no. 
I'm still not sure if I'll create a new component or just merge the code in `PerCacheInboundInvocationHandler`

In any case, I'll not add more complexity to this PR.

File: core/src/main/java/org/infinispan/commands/irac/IracUpdateKeyCommand.java [hunk 28]
```diff
@@ -0,0 +1,249 @@
...
+import org.infinispan.xsite.XSiteReplicateCommand;
+import org.infinispan.xsite.irac.IracManager;
+
+/**
+ * An update request that is sent to the remote site by {@link IracManager}.
+ *
+ * @author Pedro Ruivo
+ * @since 11.0
+ */
+public class IracUpdateKeyCommand extends XSiteReplicateCommand {
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/irac/IracUpdateKeyCommand.java:24]: sure

File: core/src/main/java/org/infinispan/xsite/irac/IracUtils.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/irac/IracUtils.java]: The guidelines for tracing is that debug can't be produced on every request for normal operation iirc. If you want to enable a specific category with TRACE at runtime, I would say go for that.

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:392]: The Command can just be null and the context can just be `ImmutableContext.INSTANCE`

File: core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java (hunk 2 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java:143]: Okay, that is what I kinda though, so we just need to make sure that we only use the BackupSender if we have a sync backup then.

File: core/src/main/java/org/infinispan/xsite/irac/IracUtils.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/IracUtils.java]: changed to trace :+1:

File: core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java (hunk 1 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/DefaultIracManager.java:388]: :+1:

File: core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java (hunk 2 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/xsite/NonTransactionalBackupInterceptor.java:143]: already changed! :+1:

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java (hunk 4 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:112]: :+1:

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java [hunk 29]
```diff
@@ -0,0 +1,307 @@
...
+public class IracEntryVersion {
+
+   private final Map<String, TopologyIracVersion> vectorClock;
+
+   public IracEntryVersion(Map<String, TopologyIracVersion> vectorClock) {
+      this.vectorClock = Objects.requireNonNull(vectorClock);
+   }
+
+   @ProtoFactory
+   static IracEntryVersion protostreamFactory(List<MapEntry> entries) {
```
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:42]: :+1:

File: core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java (hunk 6 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java:234]: :+1:

File: core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceContextInitializer.java (hunk 11 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/marshall/persistence/impl/PersistenceContextInitializer.java:51]: yes

File: core/src/main/java/org/infinispan/xsite/irac/IracUtils.java (hunk 17 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/xsite/irac/IracUtils.java:19]: it can be merged and I'll do it.

File: core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java (hunk 20 above)
Comment (pruivo) [reply core/src/test/java/org/infinispan/xsite/irac/persistence/BaseIracPersistenceTest.java:117]: -1 
That would require to remove the type `V` and use casting (and unchecked casts)
I can use `org.infinispan.persistence.PersistenceCompatibilityTest.KeyValueWrapper<K,V,T>`  since I did hit the same "problem" before.

File: core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java (hunk 10 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/AbstractIracRemoteSiteInterceptor.java:95]: done: https://issues.redhat.com/browse/ISPN-11802

**Comment by pruivo** [comment]
updated. I think I handled/replied all your comment @wburns @ryanemerson 
thanks for the review!

**Comment by ryanemerson** [comment]
Thanks @pruivo
